javascripts/content_scripts/posting.resource.js

```diff
@@ -0,0 +1,515 @@
+   * item alternately.
+   */
+  Resource.prototype.setState = function (state) {
+    this.state = state;
+    this.broadcastInternal({
+      action: 'posting/internal/stateChanged',
+      state: state
+    });
+    return true;
+  };
+
+  /**
+   * Get the state of this resource object.
+   * @return {String}
+   */
+  Resource.prototype.getState = function () {
+    return this.state;
+  };
+
+  /**
+   * Attach the resource object to the global
+   * resource pool.
+   */
+  Resource.prototype.attach = function () {
+    if (this.attached) {
+      return;
+    }
+    resource.pool.push(this);
+    this.attached = true;
+  };
+
+  /**
+   * Detach the resource object from the global
+   * resource pool.
+   */
+  Resource.prototype.detach = function () {
+    if (!this.attached) {
+      return;
+    }
+    var i;
+    for (i = 0; i < resource.pool.length; ++i) {
+      if (resource.pool[i] === this) {
+        resource.pool.splice(i, 1);
+        break;
+      }
+    }
+    this.attached = false;
+  };
+
+  /**
+   * Add a resource item into this resource.
+   * 
+   * @param {String} type The key
+   * @param {ResourceItem} instance The resource item object
+   */
+  Resource.prototype.setInstance = function (type, instance) {
+    if (this.instances[type]) {
+      // if there is already an instance using the same type name,
+      // that instance will be detached from this resource.
+      if (typeof this.instances[type].detachResource === 'function') {
+        this.instances[type].detachResource();
+      }
+    }
+    // for the resource itself, it should record the new instance
+    this.instances[type] = instance;
+    // the instance should be notified about the attach action
+    if (typeof instance.attachResource === 'function') {
+      instance.attachResource(this);
+    }
+  };
+
+  /**
+   * Get a resource item of this resource
+   * @param  {String} type item key
+   * @return {ResourceItem}
+   */
+  Resource.prototype.getInstance = function (type) {
+    return this.instances[type];
+  };
+
+  /**
+   * Remove a resource item by matching the node of the item.
+   * @param  {Node} node
+   * @return {Boolean} whether the resource item is successfully
+   * removed
+   */
+  Resource.prototype.removeInstanceByNode = function (node) {
+    var type;
+    for (type in this.instances) {
+      if (typeof this.instances[type].getNode === 'function') {
+        if (this.instances[type].getNode() === node) {
+          if (typeof this.instances[type].detachResource === 'function') {
+            this.instances[type].detachResource();
+          }
+          delete this.instances[type];
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  /**
+   * Message handler when receiving broadcast messages.
+   * This function forwards the message to all of the
+   * resource items.
+   * 
+   * @param  {String} message
+   * @param  {Function} sendResponse
```

Comment:
This is a message handler, which will forward messages to resource items attached to the resource.

For `broadcastInternal`, it acturally does not send Chrome messages. It just simulates a message. `broadcastInternal` is used for communicating between resource items like button and iframe. For example, for a Privly button, it may be in 3 state: open, close, loading. For open/close state, when user clicks it, the state toggles. For loading state, the user cannot click it. Thus, for a state controller (which handles button click and toggle state), the code will be complex and mixed together if the state controller directly listen onClick events of the button. Instead, the button class itself listens onClick event, and determine whether to send `onButtonClicked` event (message) to other components. The usage background of `broadcastInternal` is like `event`, however it can pass messages without a DOM because it reuses message interface.